File: src/ast_expr.cc
#include <ast.h>
#include "tokenizer.h"

Expr* parse_expr(Parser* p);
Expr* parse_logor(Parser* p);
Expr* parse_logand(Parser* p);
Expr* parse_bitor(Parser* p);
Expr* parse_bitxor(Parser* p);
Expr* parse_bitand(Parser* p);
Expr* parse_equ(Parser* p);
Expr* parse_comp(Parser* p);
Expr* parse_shift(Parser* p);
Expr* parse_add(Parser* p);
Expr* parse_term(Parser* p);
Expr* parse_unary(Parser* p);
Expr* parse_power(Parser* p);
Expr* parse_primary(Parser* p);
Expr* parse_literal(Parser* p);

Expr* parse_expr(Parser* p) { return parse_logor(p); }

Expr* parse_logor(Parser* p) {
    auto left = parse_logand(p);
    while (p->at(TK_LOGOR)) {
        auto op = p->advance().tk;
        auto right = parse_logand(p);

        left = new Expr(BinaryExpr{ op, left, right });
    }
    return left;
}

Expr* parse_logand(Parser* p) {
    auto left = parse_bitor(p);
    while (p->at(TK_LOGAND)) {
        auto op = p->advance().tk;
        auto right = parse_bitor(p);

        left = new Expr(BinaryExpr{ op, left, right });
    }
    return left;
}

Expr* parse_bitor(Parser* p) {
    auto left = parse_bitxor(p);
    while (p->at(TK_BITOR)) {
        auto op = p->advance().tk;
        auto right = parse_bitxor(p);

        left = new Expr(BinaryExpr{ op, left, right });
    }

    return left;
}

Expr* parse_bitxor(Parser* p) {
    auto left = parse_bitand(p);
    while (p->at(TK_BITXOR)) {
        auto op = p->advance().tk;
        auto right = parse_bitand(p);

        left = new Expr(BinaryExpr{ op, left, right });
    }

    return left;
}

Expr* parse_bitand(Parser* p) {
    auto left = parse_equ(p);
    while (p->at(TK_BITAND)) {
        auto op = p->advance().tk;
        auto right = parse_equ(p);

        left = new Expr(BinaryExpr{ op, left, right });
    }

    return left;
}

Expr* parse_equ(Parser* p) {
    auto left = parse_comp(p);
    while (p->at(TK_EQ) || p->at(TK_NEQ)) {
        auto op = p->advance().tk;
        auto right = parse_comp(p);

        left = new Expr(BinaryExpr{ op, left, right });
    }
    return left;
}

Expr* parse_comp(Parser* p) {
    auto left = parse_shift(p);
    while (p->at(TK_GT) || p->at(TK_LT) || p->at(TK_GE) || p->at(TK_LE)) {
        auto op = p->advance().tk;
        auto right = parse_shift(p);

        left = new Expr(BinaryExpr{ op, left, right });
    }
    return left;
}

Expr* parse_shift(Parser* p) {
    auto left = parse_add(p);
    while (p->at(TK_SHL) || p->at(TK_SHR)) {
        auto op = p->advance().tk;
        auto right = parse_add(p);

        left = new Expr(BinaryExpr{ op, left, right });
    }

    return left;
}

Expr* parse_add(Parser* p) {
    auto left = parse_term(p);
    while (p->at(TK_MIN) || p->at(TK_ADD)) {
        auto op = p->advance().tk;
        auto right = parse_term(p);

        left = new Expr(BinaryExpr{ op, left, right });
    }
    return left;
}

Expr* parse_term(Parser* p) {
    auto left = parse_unary(p);
    while (p->at(TK_MUL) || p->at(TK_DIV) || p->at(TK_MOD)) {
        auto op = p->advance().tk;
        auto right = parse_unary(p);

        left = new Expr(BinaryExpr{ op, left, right });
    }
    return left;
}

Expr* parse_unary(Parser* p) {
    if (p->at(TK_NOT)) {
        auto op = p->advance().tk;
        return new Expr(UnaryExpr{ op, parse_unary(p) });
    }

    if (p->at(TK_ADD) || p->at(TK_MIN)) {
        auto op = p->advance().tk;
        return new Expr(UnaryExpr{ op, parse_unary(p) });
    }

    if (p->at(TK_BNOT)) {
        auto op = p->advance().tk;
        return new Expr(UnaryExpr{ op, parse_unary(p) });
    }

    return parse_power(p);
}

Expr* parse_power(Parser* p) {
    auto left = parse_primary(p);
    if (p->at(TK_PWR)) {
        auto op = p->advance().tk;
        auto right = parse_unary(p);

        left = new Expr(BinaryExpr{ op, left, right });
    }
    return left;
}

Expr* parse_primary(Parser* p) {
    if (p->at(TK_LPAREN)) {
        p->advance();
        auto expr = parse_expr(p);

        p->expect(TK_RPAREN);
        return expr;
    }

    if (p->at(TK_IDENT)) {
        auto ident = std::get<std::string>(p->advance().val.data);
        if (p->at(TK_LPAREN)) {
            p->advance();
            std::vector<Expr*> args;
            while (!p->at(TK_RPAREN)) {
                args.push_back(parse_expr(p));

                if (p->at(TK_COMMA)) {
                    p->advance();
                }
            }
            p->advance();
            return new Expr(CallExpr{ ident, args });
        } else {
            return new Expr(IdentExpr{ ident });
        }
    }
    return parse_literal(p);
}

Expr* parse_literal(Parser* p) {
    if (p->at(TK_INT) || p->at(TK_FLOAT) ||
        p->at(TK_STRING) || p->at(TK_BOOL)) {
        return new Expr(LiteralExpr{ p->advance().val });
    }
    throw std::runtime_error(
        "unexpected token in expression: " + token_name(p->peek().tk));
}

File: src/ast_expr.rs
use crate::ast::*;

pub type ResultExpr = Result<Expr, ParserError>;

pub fn parse_expr(p: &mut Parser) -> ResultExpr {
    parse_logor(p)
}

fn parse_logor(p: &mut Parser) -> ResultExpr {
    let mut left = parse_logand(p)?;
    while p.at(Token::Logor) {
        let op = p.advance();
        let right = parse_logand(p)?;

        left = Expr::Binary {
            op,
            left: Box::new(left),
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_logand(p: &mut Parser) -> ResultExpr {
    let mut left = parse_bitor(p)?;
    while p.at(Token::Logand) {
        let op = p.advance();
        let right = parse_bitor(p)?;

        left = Expr::Binary {
            op,
            left: Box::new(left),
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_bitor(p: &mut Parser) -> ResultExpr {
    let mut left = parse_bitxor(p)?;
    while p.at(Token::Bitor) {
        let op = p.advance();
        let right = parse_bitxor(p)?;

        left = Expr::Binary {
            op,
            left: Box::new(left),
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_bitxor(p: &mut Parser) -> ResultExpr {
    let mut left = parse_bitand(p)?;
    while p.at(Token::Bitand) {
        let op = p.advance();
        let right = parse_bitand(p)?;

        left = Expr::Binary {
            op,
            left: Box::new(left),
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_bitand(p: &mut Parser) -> ResultExpr {
    let mut left = parse_equ(p)?;
    while p.at(Token::Bitand) {
        let op = p.advance();
        let right = parse_equ(p)?;

        left = Expr::Binary {
            op,
            left: Box::new(left),
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_equ(p: &mut Parser) -> ResultExpr {
    let mut left = parse_comp(p)?;
    while p.at(Token::Eq) || p.at(Token::Neq) {
        let op = p.advance();
        let right = parse_comp(p)?;

        left = Expr::Binary {
            op,
            left: Box::new(left),
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_comp(p: &mut Parser) -> ResultExpr {
    let mut left = parse_shift(p)?;
    while p.at(Token::Gt) || p.at(Token::Lt) || p.at(Token::Ge) || p.at(Token::Le) {
        let op = p.advance();
        let right = parse_shift(p)?;

        left = Expr::Binary {
            op,
            left: Box::new(left),
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_shift(p: &mut Parser) -> ResultExpr {
    let mut left = parse_add(p)?;
    while p.at(Token::Shl) || p.at(Token::Shr) {
        let op = p.advance();
        let right = parse_add(p)?;

        left = Expr::Binary {
            op,
            left: Box::new(left),
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_add(p: &mut Parser) -> ResultExpr {
    let mut left = parse_term(p)?;
    while p.at(Token::Min) || p.at(Token::Add) {
        let op = p.advance();
        let right = parse_term(p)?;

        left = Expr::Binary {
            op,
            left: Box::new(left),
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_term(p: &mut Parser) -> ResultExpr {
    let mut left = parse_unary(p)?;
    while p.at(Token::Mul) || p.at(Token::Div) || p.at(Token::Mod) {
        let op = p.advance();
        let right = parse_unary(p)?;

        left = Expr::Binary {
            op,
            left: Box::new(left),
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_unary(p: &mut Parser) -> ResultExpr {
    if p.at(Token::Not) || p.at(Token::Add) || p.at(Token::Min) || p.at(Token::Bnot) {
        let op = p.advance();
        return Ok(Expr::Unary {
            op,
            arg: Box::new(parse_unary(p)?),
        });
    }

    parse_power(p)
}

fn parse_power(p: &mut Parser) -> ResultExpr {
    let mut left = parse_primary(p)?;
    if p.at(Token::Pwr) {
        let op = p.advance();
        let right = parse_primary(p)?;

        left = Expr::Binary {
            op,
            left: Box::new(left),
            right: Box::new(right),
        };
    }

    Ok(left)
}

fn parse_primary(p: &mut Parser) -> ResultExpr {
    if p.at(Token::Lparen) {
        p.advance();
        let expr = parse_expr(p);

        p.expect(Token::Rparen)?;
        return expr;
    }

    let mut base: Expr;

    if let Token::Ident(id) = p.peek() {
        p.advance();
        if p.at(Token::Lparen) {
            p.advance();
            let mut args = Vec::<Expr>::new();
            while !p.at(Token::Rparen) {
                args.push(parse_expr(p)?);
                if p.at(Token::Comma) {
                    p.advance();
                }
            }

            p.advance();
            base = Expr::Call { ident: id, args };
        } else {
            base = Expr::Ident { ident: id };
        }
    } else {
        base = parse_literal(p)?;
    }

    while p.at(Token::Lbrack) {
        p.advance();
        let idx = parse_expr(p)?;
        p.expect(Token::Rbrack)?;
        base = Expr::Index {
            expr: Box::new(base),
            idx: Box::new(idx),
        };
    }

    Ok(base)
}

fn parse_literal(p: &mut Parser) -> ResultExpr {
    if p.at(Token::Lbrace) {
        return parse_aliteral(p);
    }

    if let Token::Int(i) = p.peek() {
        p.advance();
        Ok(Expr::Literal { val: Value::Int(i) })
    } else if let Token::Float(f) = p.peek() {
        p.advance();
        Ok(Expr::Literal {
            val: Value::Float(f),
        })
    } else if let Token::String(s) = p.peek() {
        p.advance();
        Ok(Expr::Literal {
            val: Value::String(s),
        })
    } else if let Token::Bool(b) = p.peek() {
        p.advance();
        Ok(Expr::Literal {
            val: Value::Bool(b),
        })
    } else {
        Err(ParserError::UnexpectedToken)
    }
}

fn parse_aliteral(p: &mut Parser) -> ResultExpr {
    p.expect(Token::Lbrace)?;
    let mut mbrs = Vec::<Expr>::new();

    while !p.at(Token::Rbrace) {
        mbrs.push(parse_expr(p)?);
        if p.at(Token::Comma) {
            p.advance();
        }
    }

    p.advance();

    Ok(Expr::ArrayLiteral { members: mbrs })
}
